### PR TITLE
Cleanse Service Provider Interface (aka Plugins)

### PR DESCRIPTION
**Pull request by sebastianv1**
* Adds support for a plugin interface that developers can use to hook into the object graph.
* Interface is currently readonly.
* RFC: #118

### CODE REVIEW

**Comment by holmes** [comment]
Looks good, but why not convert the `ValidationVisitor` to a `CleanseBindingPlugin` to prove that it works?

**Comment by sebastianv1** [comment]
Could be an option, but I do like that it's part of the core framework and not just a plugin. Could be worth migrating one day.

**Comment by holmes** [comment]
It'd be nice to have something as a proof of concept that this works. If
you don't want to convert that one I'd strongly encourage you to create one
in the demo.

A working example is the most likely way to get people to build on this
feature.

**Comment by sebastianv1** [comment]
That's a good point. I'll work on an example impl with some tests.

**Comment by sebastianv1** [comment]
@holmes I've added a simple unit test. My thinking is to create a separate framework for the plugin samples that can be consumed independently of `Cleanse`. This would involve adding a new podspec (named `CleansePlugins` for instance) that would depend on `Cleanse`. In order for this to happen, I'd rather land this and have `CleansePlugins` point at the master branch of Cleanse instead of this development branc.

File: CleanseTests/SPITests.swift [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    
+    struct MyPlugin: CleanseBindingPlugin {
+        let expectation: XCTestExpectation
+        func visit(root: ComponentBinding, errorReporter: CleanseErrorReporter) {
+            expectation.fulfill()
+        }
+    }
+    
+    func testPluginRuns() {
+        let expection = XCTestExpectation(description: "Plugin should run")
```
Comment (holmes) [inline CleanseTests/SPITests.swift]: spelling: `expectation`

**Pull request merged**